### PR TITLE
MGDOBR-1241 Automatically update Shard router canonical hostname

### PR DESCRIPTION
**Pull request by sutaakar**
Signed-off-by: Karel Suta <ksuta@redhat.com>

[MGDOBR-1241](https://issues.redhat.com/browse/MGDOBR-1241)

## PR owner checklist

Please make sure that your PR meets the following requirements:

- [x] Your code is properly formatted according to [this configuration](https://github.com/kiegroup/kogito-runtimes/tree/main/kogito-build/kogito-ide-config).  
  *Run `mvn clean verify -DskipTests` so that imports are correctly set.*
- [x] Your commit messages are clear and reference the JIRA issue e.g: "[MGDOBR-1] - $clear_explanation_of_what_you_did"
- [x] The layers in the `kustomize` folder have been updated accordingly.
- [x] All new functionality is tested
- [x] Pull Request title is properly formatted: `MGDOBR-XYZ Subject`
- [x] Pull Request contains link to the JIRA issue
- [x] Pull Request contains link to any dependent or related Pull Request
- [x] Pull Request contains description of the issue
- [x] Pull Request does not include fixes for issues other than the main ticket

## PR reviewer(s) checklist

- [ ] Check if code and Github action workflows have been modified and if the modification is safe
- [ ] If the modification is safe, add the `safe to test` label

<details>
<summary>
How to trigger pipelines and use the bots:
</summary>

- <b>Run the end to end pipeline</b>  
  Annotate the pull request with the label: `safe to test`. If you want to run the pipeline again, remove and add it again.

- <b>Rebase the pull request</b>  
  Comment with: `/rebase`.
- <b>Deploy BOT</b>
  Comment with `/deploy <dev,stable> when the PR has been merged to deploy to a target environment.

</details>

### CODE REVIEW

File: manager-parent/manager-core/src/main/java/com/redhat/service/smartevents/manager/core/services/ShardServiceImpl.java [hunk 1]
```diff
@@ -30,7 +35,25 @@ public void init() {
...
+            if (!shardRouterCanonicalHostnameFromProperties.equals(routerCanonicalHostname)) {
+                LOGGER.warn("Shard router canonical hostname '{}' doesn't match expected value '{}' from properties file, replacing hostname to match properties.", routerCanonicalHostname,
+                        shardRouterCanonicalHostnameFromProperties);
+                shard.setRouterCanonicalHostname(shardRouterCanonicalHostnameFromProperties);
+                persist(shard);
+            }
+        }
+    }
+
+    @Transactional
```
Comment (robpblake) [inline manager-parent/manager-core/src/main/java/com/redhat/service/smartevents/manager/core/services/ShardServiceImpl.java:54]: @sutaakar If you move the `@Transactional` annotation onto the `init` method then you can delete this method.

The `ShardDAO.listAll()` method will load the `Shard` entity into the JPA `PersistenceContext` (move the entity into the "managed" lifecycle state). Your call to `Shard.setRouterCanonicalHostname` will mark the entity as "dirty" and then Hibernate will automatically flush the changes to the DB when the transaction commits. 

There's a high-level overview here of entity lifecycles and the PersistenceContext: https://www.baeldung.com/hibernate-entity-lifecycle#:~:text=Every%20Hibernate%20entity%20naturally%20has,able%20to%20use%20Hibernate%20properly.

cheers,

Rob

**Review by robpblake — COMMENTED** [review]
@sutaakar Minor feedback from me, but otherwise looks good.

cheers,
Rob

File: manager-parent/manager-core/src/main/java/com/redhat/service/smartevents/manager/core/services/ShardServiceImpl.java (hunk 1 above)
Comment (sutaakar) [reply manager-parent/manager-core/src/main/java/com/redhat/service/smartevents/manager/core/services/ShardServiceImpl.java:54]: @robpblake Unfortunately `@Transactional` is not compatible with `@PostConstruct`. As a result when both annotations are used on same method then transaction doesn't seem to be opened - changes are not persisted.

Comment (robpblake) [reply manager-parent/manager-core/src/main/java/com/redhat/service/smartevents/manager/core/services/ShardServiceImpl.java:54]: @sutaakar hmm...shame. Thanks for trying it though.

**Review by robpblake — APPROVED** [review]

File: manager-parent/manager-core/src/main/java/com/redhat/service/smartevents/manager/core/services/ShardServiceImpl.java (hunk 1 above)
Comment (lucamolteni) [reply manager-parent/manager-core/src/main/java/com/redhat/service/smartevents/manager/core/services/ShardServiceImpl.java:54]: we could do a JPA update then `where name != shard.getRouterCanonicalHostname()`

There's also `update` on Panache but I've never used it

**Review by lucamolteni — APPROVED** [review]
Make a try and see if you can shorten significantly the method with the update otherwise it's fine by me

**Pull request merged**